### DISCUSSION TITLE
Replace one or multiple Arguments with a new Object

### DISCUSSION THREAD

**Discussion by alexander-schranz** [rectorphp/rector#6791] [Q&A]
I'm trying to refractor service to message bus, I already was able renaming the the method and call the service bus. What I'm stuck with is to replace an argument with a wrapped new constructor. 

```diff
-$service->save($one);
+$messageBus->dispatch(new CreateMediaMessage($one);
```

Additional it would be great if possible to depend on the argument numbers which message should be instance and change the argument orders to the new method e.g.:

```diff
-$service->save($one, $two);
+$messageBus->dispatch(new ModifyMediaMessage($two, $one);
```

I did not yet found a refractor rule to achieve this. But I think I did just oversee the correct rule. Would be great if somebody could show me the right way to achieve this.

**Comment by TomasVotruba** [comment]
Hi, thanks for sharing. There is no rule to handle this, as it's very specific job not typical for BC breaks in open source:

* rename variable
* rename method
* replace argument
* replace variable with value object
* reuse original arguments as new initializers

Yet, there is a way to handle this. Create a custom rule, hook on MethodCall and change what you need.

**Reply by alexander-schranz** [reply]
Thank you for your fast response. I will try to have a look maybe I find a general rule which I can contribute here. Example rule with:

```php
                new MethodCallReplaceArguments(
                    'Sulu\\Bundle\\MediaBundle\\Media\\Manager\\MediaManagerInterface',
                    'save',
                    1,
                    'new CreateMediaMessage(%1)'
                ),
                new MethodCallReplaceArguments(
                    'Sulu\\Bundle\\MediaBundle\\Media\\Manager\\MediaManagerInterface',
                    'save',
                    2,
                    'new ModifyMediaMessage(%1, %2)'
                ),
                // with more arguments
                new MethodCallReplaceArguments(
                    'Sulu\\Bundle\\MediaBundle\\Media\\Manager\\MediaManagerInterface',
                    'save',
                    2,
                    'new ModifyMediaMessage(%1, %2), [new EnabledFlushStamp()]'
                ),
```

**Reply by TomasVotruba** [reply]
> I will try to have a look maybe I find a general rule which I can contribute here.

That could be useful :+1: 
Instead of magic string expressions, the input should be as strict as possible:

```php
new WrapMethodCallArgToValueObject(
                    'SomeService',
                    'save',
                    'CreateMediaMessage'
                ),
```

```diff
-$someService->save(1, 2, 3);
+$someService->save(new CreateMediaMessage(1, 2, 3));
```

This way we avoid parsing-string bugs and input assumptions.

**Reply by alexander-schranz** [reply]
Thanks for the Input. I hope I will find this week some time to have a look at it. I would also think the position and the used values should be optional configureable:

```php
new WrapMethodCallArgToValueObject(
    'SomeService',
    'save',
    'CreateMediaMessage',
    0, // position of the value object
    [1, 0] // arguments used for the value object (others will be unmodified)
),
```

```diff
-$someService->save(1, 2, 3);
+$someService->save(new CreateMediaMessage(2, 1), 3);
```
